src/ray/raylet/worker_pool_test.cc

```diff
@@ -1233,6 +1256,54 @@ TEST_F(WorkerPoolTest, StartWorkWithDifferentShimPid) {
   ASSERT_TRUE(worker_pool_->PopWorker(java_task_spec) != nullptr);
 }
 
+TEST_F(WorkerPoolTest, NoSpuriousWorkerStartupDuringEnvInstall) {
+  std::vector<std::string> agent_commands = {};
+  const NodeID node_id = NodeID::FromRandom();
+  auto options = AgentManager::Options({node_id, agent_commands});
+  auto agent_manager = std::make_shared<MockAgentManager>(
+      std::move(options),
+      /*delay_executor=*/
+      [this](std::function<void()> task, uint32_t delay_ms) {
+        return execute_after(io_service_, task, delay_ms);
+      },
+      /*runtime_env_agent_factory=*/
+      [](const std::string &ip_address, int port) {
+        return std::shared_ptr<rpc::RuntimeEnvAgentClientInterface>(
+            new MockRuntimeEnvAgentClient());
+      },
+      false);
+  worker_pool_->SetAgentManager(agent_manager);
+
+  ASSERT_EQ(worker_pool_->GetProcessSize(), 0);
+  const auto normal_task_spec =
+      ExampleTaskSpec(ActorID::Nil(), Language::PYTHON, JOB_ID, ActorID::Nil(),
+                      /*dynamic_options=*/{}, TaskID::Nil(), "mock_runtime_env_1");
+
+  // Pop worker for a task with runtime env.
+  auto popped_worker = worker_pool_->PopWorker(normal_task_spec);
+
+  // No idle workers.
+  ASSERT_EQ(popped_worker, nullptr);
+
+  // Worker does not start because CreateRuntimeEnvCallback has not yet been invoked.
+  ASSERT_EQ(worker_pool_->GetProcessSize(), 0);
+
+  // Simulate the following situation: before runtime env is done installing,
+  // PopWorker is called several more times for the same task.
+  for (int i = 0; i < 10; i++) {
+    worker_pool_->PopWorker(normal_task_spec);
+  }
+
+  // AgentManager::CreateRuntimeEnv() should have only been called once.
+  ASSERT_EQ(agent_manager->queued_callbacks.size(), 1);
+
+  // Finish installing runtime env, call CreateRuntimeEnvCallback.
+  agent_manager->PopAndInvokeCallback();
+
+  // Only one worker process is started.
+  ASSERT_EQ(worker_pool_->GetProcessSize(), 1);
+}
```

Comment:
We should also check that after this we can create a second worker with the same env.

You might consider adding a test that multiple RPCs will or will not be sent in parallel depending on the resolution from the comment above about this approach.